tools/query-screen.py

```diff
@@ -0,0 +1,114 @@
+{
+    "user":<username>,
+    "screens": [
+        {
+        "idNum":<id number>,
+        "idName":<id name>,
+        "state":<state>
+        },
+        ...
+    ]
+}
+
+To create a detached screen, use `-dm`:
+`screen -dm`
+
+To create a named screen, use `-S`:
+`screen -dm -S <screenName>`
+ex `screen -dm -S myScreen`
+
+To list the currently running screen sessions, use:
+`screen -ls`
+
+To list the currently running screen sessions of a particular ie different user, use:
+`sudo -u <username> screen -ls`
+ex `sudo -u $(whoami) screen -ls`
+
+To attach to a detached screen, refer to the screen using the id number or the name.
+`screen -r <screenId|screenName>`
+ex `screen -r myScreen`
+
+In an attached screen session, you may press `<ctrl + a>`, then `:` to enter 'command mode'.
+
+To detach from an attached screen session, enter command mode, and press `<d>` to detach.
+To scroll in an attached screen session, enter command mode, and press `<esc>`.
+Then, you may scroll. To resume following the most recent output, press `<esc>` again.
+
+To issue a command on a detached screen, use `-X` and the command you want to enter, `quit`:
+`screen -S <screenName> -X <command>`
+ex `screen -S myScreen -X quit`
+
+To exit (terminate) an attached screen session, enter command mode, and type `quit`, 
+then press `<enter>`.
+
+See `man screen` for more information.
+'''
+
+import argparse
+import subprocess
+import re
+import json
+
+PARSER = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawTextHelpFormatter)
+PARSER.add_argument('--user',
+                    action="store",
+                    type=str,
+                    default="root",
+                    help="User to check for running screen processes."
+                    )
+PARSER.add_argument('--indent',
+                    action="store",
+                    type=int,
+                    default=None,
+                    help="Indent level for json output formatting."
+                    )
+ARGS = PARSER.parse_args()
+
+ret = {
+    "user" : ARGS.user,
+    "screens" : []
+}
+
+# process line of interest (regex match) from output of `screen -ls` command
+# ex:
+# $ screen -ls
+# There are screens on:
+#     3669.myScreen      (Detached)
+def processScreenOutputLine(line):
+    ret = {}
+
+    fields = line.split('\t')
+    id = fields[0].split('.', 1)
+    ret["idNum"] = id[0]
+    ret["idName"] = id[1]
+    ret["state"] = fields[-1]
+    
+    return ret
+
+output = ""
+try:
+    # -A exits w error instead of asking password
+    output = subprocess.check_output(["sudo", "-Au" + ARGS.user, "screen", "-ls"], universal_newlines=True)
+
+except subprocess.CalledProcessError as e:
+    if e.returncode is 1:
+        output = e.output # screen v4.2.x always gives exit code of 1 for 'screen -ls'
+
+finally:
+    # match 1~many digits, followed by one dot '.' character, then whatever follows until we hit a newline
+    regex = re.compile("[0-9]+\..*")
```

Comment:
I think that we could use a regex named capturing group to figure out everything for us.

Reply:
I don't know how to do this. Can you show me what you mean while I read the doc?

Reply:
"[0-9]{3}\.(?<name>\S*)"
https://docs.python.org/2/library/re.html

Reply:
@patrickelectric @Williangalvani is this what you mean? regexr.com/44qov

This takes longer to create and also read/interpret than the original code, I think. Do you guys think it's better to replace this? We _can_ but should we?

Reply:
> "[0-9]{3}\.(?<name>\S*)"
https://docs.python.org/2/library/re.html

I still do not know what you mean:
> I think that we could use a regex named capturing group to figure out everything for us.

Reply:
I was thinking about this:
https://regex101.com/r/cOgkdu/1

Reply:
@Williangalvani different versions of screen have different formats, the pi does not show the date/time. So yours does not work for all of them.

Is this change needed for the pull request?

Reply:
Also @Williangalvani the screen name may have space characters or period characters which are not captured with \w.

Reply:
this seems to catch all: https://regex101.com/r/cOgkdu/3
I havent worked with groups in regexes yet, so I'm not sure how straight forward it it to implement like this, but I think it would probably be shorter.
I think it is worth taking a look at this solution, but feel free to keep it as you wish.

Reply:
The PR works, it's not overly long or complicated as is, and shorter code is not necessarily better code. I agree it's worth looking on better regexes for this and similar in the future, but I've already spent some hours today reading about it before I even start to change the python code. To me it's something that's better to follow up with after the PR.

Reply:
sure, no problem.

Reply:
Well I couldn't stop until I had it figured out, I updated it.

Reply:
here we are: https://regex101.com/r/LrtbyA/1

Reply:
cool!
I expected it to return a dict using the group names or something, [but apparently it doesn't](https://stackoverflow.com/questions/11103856/re-findall-which-returns-a-dict-of-named-capturing-groups)